### PR TITLE
Add Phala runtime

### PR DESCRIPTION
**Pull request by jasl**

### CODE REVIEW

**Comment by jasl** [comment]
@h4x3rotab please help to:
- Generate a set of new secret keys for Phala runtime
- Set correct chainspec parameters for Phala

**Comment by h4x3rotab** [comment]
Addressed at: https://github.com/Phala-Network/khala-parachain/pull/48

**Comment by jasl** [comment]
@h4x3rotab still lacking set correct chainspec parameters for Phala, but the PR is ready for merge, I've tested in polkadot-launch

File: node/src/chain_spec/phala.rs [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+            phala_genesis(
+                genesis_info.root_key,
+                genesis_info.initial_authorities,
+                genesis_info.technical_committee,
+                genesis_info
+                    .endowed_accounts
+                    .into_iter()
+                    .map(|(k, amount)| (k, u128::from_str(&amount).expect("Bad amount; qed.")))
+                    .collect(),
+                2004u32.into(),
```
Comment (h4x3rotab) [inline node/src/chain_spec/phala.rs:161]:
```suggestion
                2035u32.into(),
```

File: node/src/chain_spec/phala.rs [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+            )
+        },
+        Vec::new(),
+        None,
+        Some("phala"),
+        chain_properties(),
+        Extensions {
+            relay_chain: "kusama".into(),
+            para_id: 2004,
```
Comment (h4x3rotab) [inline node/src/chain_spec/phala.rs:170]:
```suggestion
            para_id: 2035,
```

Comment (jasl) [reply node/src/chain_spec/phala.rs]: Changed

File: node/src/chain_spec/phala.rs (hunk 1 above)
Comment (jasl) [reply node/src/chain_spec/phala.rs]: Changed

File: node/src/service/mod.rs [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+    };
+
+    Ok(params)
+}
+
+/// Start a node with the given parachain `Configuration` and relay chain `Configuration`.
+///
+/// This is the actual implementation that is abstract over the executor and the runtime api.
+#[sc_tracing::logging::prefix_logs_with("Parachain")]
+async fn start_node_impl<RuntimeApi, Executor, RB, BIQ, BIC>(
```
Comment (h4x3rotab) [inline node/src/service/mod.rs:160]: It looks like we already have start_node_impl for khala, phala, and thala. Where is this function used?

File: runtime/phala/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,1214 @@
...
+    }
+}
+
+/// This runtime version.
+#[sp_version::runtime_version]
+pub const VERSION: RuntimeVersion = RuntimeVersion {
+    spec_name: create_runtime_str!("phala"),
+    impl_name: create_runtime_str!("phala"),
+    authoring_version: 1,
+    spec_version: 1090,
```
Comment (h4x3rotab) [inline runtime/phala/src/lib.rs:122]: Will we sync the spec version between khala and phala?

File: runtime/phala/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,1214 @@
...
+    EnsureRoot<AccountId>,
+    pallet_collective::EnsureProportionMoreThan<_1, _2, AccountId, CouncilCollective>,
+>;
+
+construct_runtime! {
+    pub enum Runtime where
+        Block = Block,
+        NodeBlock = opaque::Block,
+        UncheckedExtrinsic = UncheckedExtrinsic,
+    {
```
Comment (h4x3rotab) [inline runtime/phala/src/lib.rs:179]: I haven't check. Are the pallet indices in sync with Khala?

File: runtime/phala/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,1214 @@
...
+        Lottery: pallet_lottery::{Pallet, Call, Storage, Event<T>} = 65,
+        TechnicalCommittee: pallet_collective::<Instance2>::{Pallet, Call, Storage, Origin<T>, Event<T>, Config<T>} = 66,
+        TechnicalMembership: pallet_membership::<Instance1>::{Pallet, Call, Storage, Event<T>, Config<T>} = 67,
+        PhragmenElection: pallet_elections_phragmen::{Pallet, Call, Storage, Event<T>, Config<T>} = 68,
+        Tips: pallet_tips::{Pallet, Call, Storage, Event<T>} = 69,
+
+        // Main, starts from 80
+
+        // TODO: Remove sudo in the future
+        Sudo: pallet_sudo::{Pallet, Call, Storage, Config<T>, Event<T>} = 255,
```
Comment (h4x3rotab) [inline runtime/phala/src/lib.rs:221]: Not sure if 255 is supported

File: runtime/phala/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,1214 @@
...
+        Scheduler: pallet_scheduler::{Pallet, Call, Storage, Event<T>} = 7,
+
+        // Parachain staff
+        ParachainInfo: pallet_parachain_info::{Pallet, Storage, Config} = 20,
+        ParachainSystem: cumulus_pallet_parachain_system::{Pallet, Call, Config, Storage, Inherent, Event<T>, ValidateUnsigned} = 21,
+
+        // Monetary stuff
+        Balances: pallet_balances::{Pallet, Call, Storage, Config<T>, Event<T>} = 40,
+        TransactionPayment: pallet_transaction_payment::{Pallet, Storage} = 41,
+        Assets: pallet_assets::{Pallet, Call, Storage, Event<T>} = 42,
```
Comment (h4x3rotab) [inline runtime/phala/src/lib.rs:197]: If we put Assets at 42, we should also move Assets in thala to the same slot

File: runtime/phala/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,1214 @@
...
+    type Currency = Balances;
+    type DepositBase = DepositBase;
+    type DepositFactor = DepositFactor;
+    type MaxSignatories = MaxSignatories;
+    type WeightInfo = pallet_multisig::weights::SubstrateWeight<Runtime>;
+}
+
+parameter_types! {
+    // One storage item; key size 32, value size 8; .
+    pub const ProxyDepositBase: Balance = deposit(1, 40);
```
Comment (h4x3rotab) [inline runtime/phala/src/lib.rs:356]: Didn't check, but if it's in sync with Khala, we should be safe

File: runtime/phala/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,1214 @@
...
+    type MaxPending = MaxPending;
+    type CallHasher = Hasher;
+    type AnnouncementDepositBase = AnnouncementDepositBase;
+    type AnnouncementDepositFactor = AnnouncementDepositFactor;
+}
+
+parameter_types! {
+    pub MaximumSchedulerWeight: Weight = Perbill::from_percent(80) *
+        RuntimeBlockWeights::get().max_block;
+    pub const MaxScheduledPerBlock: u32 = 50;
```
Comment (h4x3rotab) [inline runtime/phala/src/lib.rs:481]: Sounds too large? But if it's in sync, we should be fine

File: runtime/phala/src/lib.rs (hunk 4 above)
Comment (jasl) [reply runtime/phala/src/lib.rs:122]: I think yes, this follows Polkadot's rule

File: node/src/service/mod.rs (hunk 3 above)
Comment (jasl) [reply node/src/service/mod.rs:160]: most are the same, but Phala doesn't have Phala pallet, so they have slightly different:
- generic constraint, no <https://github.com/Phala-Network/khala-parachain/pull/46/files#diff-daced65c5a2c24e21e748746a2501eb6f6006d23e3d5fc15804227d84496f57aR186>
- rpc expose

File: runtime/phala/src/lib.rs (hunk 5 above)
Comment (jasl) [reply runtime/phala/src/lib.rs:179]: yes, copied from Khala

File: runtime/phala/src/lib.rs (hunk 6 above)
Comment (jasl) [reply runtime/phala/src/lib.rs:221]: learned from Acala <https://github.com/AcalaNetwork/Acala/blob/master/runtime/acala/src/lib.rs#L1833>

File: runtime/phala/src/lib.rs (hunk 7 above)
Comment (jasl) [reply runtime/phala/src/lib.rs:197]: Thala also 42 <https://github.com/AcalaNetwork/Acala/blob/master/runtime/acala/src/lib.rs#L1833>

File: runtime/phala/src/lib.rs (hunk 8 above)
Comment (jasl) [reply runtime/phala/src/lib.rs:356]: same <https://github.com/Phala-Network/khala-parachain/blob/main/runtime/khala/src/lib.rs#L358>

File: runtime/phala/src/lib.rs (hunk 9 above)
Comment (jasl) [reply runtime/phala/src/lib.rs:481]: same <https://github.com/Phala-Network/khala-parachain/blob/main/runtime/khala/src/lib.rs#L492>

File: node/src/service/mod.rs (hunk 3 above)
Comment (h4x3rotab) [reply node/src/service/mod.rs:160]: okay

**Pull request merged**